fiftyone/core/dataset.py

```diff
@@ -652,42 +674,60 @@ def _add_samples_batch(self, samples, expand_schema):
 
         return [str(d["_id"]) for d in dicts]
 
-    def merge_samples(self, samples, overwrite=False):
+    def merge_samples(
+        self, samples, overwrite=False, key_field="filepath", key_fcn=None
+    ):
         """Merges the contents of the given samples into the dataset.
 
-        Input samples whose ``filepath`` matches an existing ``filepath`` are
-        merged, and samples with new ``filepath`` values are added.
+        By default, samples with the same absolute ``filepath`` are merged.
+
+        You can customize this behavior via the ``key_field`` and ``key_fcn``
+        parameters. For example, you could set
+        ``key_fcn = lambda k: os.path.basename(k)`` to merge samples with the
+        same base filename.
 
         Args:
             samples: an iterable of :class:`fiftyone.core.sample.Sample`
                 instances. For example, ``samples`` may be a :class:`Dataset`
                 or a :class:`fiftyone.core.views.DatasetView`
             overwrite (False): whether to overwrite (True) or skip (False)
                 existing sample fields
+            key_field ("filepath"): the sample field to use to decide whether
+                to join with an existing sample
+            key_fcn (None): a function to apply to ``key_field`` to generate
+                the comparator used to decide if two samples should be merged
         """
-        existing_schema = self.get_field_schema()
-        filepath_map = {s.filepath: s.id for s in self.select_fields()}
-
-        for new_sample in samples:
-            if new_sample.filepath in filepath_map:
-                existing_sample = self[filepath_map[new_sample.filepath]]
+        if key_fcn is None:
+            key_fcn = lambda k: k
 
-                for name, value in new_sample.iter_fields():
-                    if name == "filepath":
-                        continue
-
-                    if (
-                        not overwrite
-                        and name in existing_schema
-                        and existing_sample[name] is not None
-                    ):
-                        continue
+        existing_schema = self.get_field_schema()
 
-                    existing_sample[name] = value
+        id_map = {}
+        for sample in self.select_fields(key_field):
```

Comment:
Not ideal to create a dataset large map, but it would also take quite a large dataset for this to be a problem.

Ideally this merging could be done purely within mongodb.

Reply:
Yeah it's just a `str -> str` map of size `O(samples)`, so not too bad.

Would've loved to implement this method directly in the DB, but don't have the MongoDB chops to do it yet 🤓